substrate/client/executor/src/wasm_runtime.rs

```diff
@@ -293,34 +297,207 @@ pub fn create_wasm_runtime_with_code<H>(
-						deterministic_stack_limit: None,
-						canonicalize_nans: false,
-						parallel_compilation: true,
-						wasm_multi_value: false,
-						wasm_bulk_memory: false,
-						wasm_reference_types: false,
-						wasm_simd: false,
+		WasmExecutionMethod::Compiled { instantiation_strategy } => {
+			let semantics = sc_executor_wasmtime::Semantics {
+				heap_alloc_strategy,
+				instantiation_strategy,
+				deterministic_stack_limit: None,
+				canonicalize_nans: false,
+				parallel_compilation: true,
+				wasm_multi_value: false,
+				wasm_bulk_memory: false,
+				wasm_reference_types: false,
+				wasm_simd: false,
+			};
+			if let Some(wasmtime_precompiled_dir) = wasmtime_precompiled_path {
+				if !wasmtime_precompiled_dir.is_dir() {
+					return Err(WasmError::Instantiation(format!(
+						"--wasmtime-precompiled is not a directory: {}",
+						wasmtime_precompiled_dir.display()
+					)))
+				}
+				let handle_err = |e: std::io::Error| -> WasmError {
+					return WasmError::Instantiation(format!(
+						"Io error when loading wasmtime precompiled folder ({}): {}",
+						wasmtime_precompiled_dir.display(),
+						e
+					))
+				};
+				let mut maybe_compiled_artifact = None;
+
+				let artifact_version =
+					compute_artifact_version(allow_missing_func_imports, code_hash, &semantics);
+				log::debug!(
+					target: "wasmtime-runtime",
+					"Searching for wasm hash: {}",
+					artifact_version.clone()
+				);
+
+				for entry in std::fs::read_dir(wasmtime_precompiled_dir).map_err(handle_err)? {
+					let entry = entry.map_err(handle_err)?;
+					if let Some(file_name) = entry.file_name().to_str() {
+						// We check that the artifact was generated for this specific artifact
+						// version and with the same wasm interface version and configuration.
+						if file_name.contains(&artifact_version.clone()) {
+							log::info!(
+								target: "wasmtime-runtime",
+								"Found precompiled wasm: {}",
+								file_name
+							);
+							// We change the version check strategy to make sure that the file
+							// content was serialized with the exact same config as well
+							maybe_compiled_artifact = Some((
+								entry.path(),
+								sc_executor_wasmtime::ModuleVersionStrategy::Custom(
+									artifact_version.clone(),
+								),
+							));
+						}
+					} else {
+						return Err(WasmError::Instantiation(
+							"wasmtime precompiled folder contain a file with invalid utf8 name"
+								.to_owned(),
+						))
+					}
+				}
+
+				if let Some((compiled_artifact_path, module_version_strategy)) =
+					maybe_compiled_artifact
+				{
+					// # Safety
+					//
+					// The file name of the artifact was checked before,
+					// so if the user has not renamed nor modified the file,
+					// it's certain that the file has been generated by
+					// `prepare_runtime_artifact` and with the same wasmtime
+					// version and configuration.
+					unsafe {
+						sc_executor_wasmtime::create_runtime_from_artifact::<H>(
```

Comment:
```suggestion
						return sc_executor_wasmtime::create_runtime_from_artifact::<H>(
```

If you make this an early return, below we can move the `create_runtime` out of the if to only have it once.